content/advent-2015/test2doc-generate-your-api-docs.md

```diff
@@ -0,0 +1,294 @@
++++
+author = ["Sarah Adams"]
+date = "2015-12-14T22:04:31-08:00"
+series = ["Advent 2015"]
+title = "test2doc: Generate Your API Docs"
+
++++
+
+API docs are important. We all know that.
+They are also painful and tedious to maintain.
+When your docs aren't accurate, you get more questions. And everyone 
+loses time.
+
+I've always thought of unit tests as a great source of documentation. 
+But the non-Go engineers consuming my API don't often agree.
+
+Yet most of the same information is there:
+
+- request format:
+    - HTTP method
+    - URI
+    - query params
+    - request headers
+    - request body
+- response format:
+    - status code
+    - response headers
+    - response body
+
+<br>
+All that's missing from this list is the 
+**high-level descriptions**, the **context**.
+
+Enter *Go doc*.<br>
+We can find descriptions of an HTTP handler in its Go doc string, eg:
+```go
+// GetWidget retrieves a single Widget
+func GetWidget(w http.ResponseWriter, req *http.Request) {
+    // ...
+}
+```
+<br>
+And thus began **test2doc** - automatically generate complete API 
+documentation from your existing Go unit tests + Go doc strings.
+
+
+## Example
+
+Given an HTTP handler func, eg.:
+
+```go
+// GetWidget retrieves a single Widget
+func GetWidget(w http.ResponseWriter, req *http.Request) {
+    // ...
+}
+```
+
+And a test for this handler func, eg.:
+
+```go
+func TestGetWidget(t *testing.T) {
+    urlPath := fmt.Sprintf("/widgets/%d", 2)
+
+    resp, err := http.Get(server.URL + urlPath)
+    // assert all the things...
+}
+```
+
+Test2doc will automatically generate markdown documentation for this 
+endpoint in the [API 
+Blueprint](https://github.com/apiaryio/api-blueprint/blob/master/API%20B
+lueprint%20Specification.md) format as your tests run, like so:
+
+```
+# Group widgets
+
+## /widgets/{id}
+
++ Parameters
+    + id: `2` (number)
+
+### Get Widget [GET]
+retrieves a single Widget
+
++ Response 200 
+
+    + Body
+
+            {
+                "Id": 2,
+                "Name": "Pencil",
+                "Role": "Utensil"
+            }        
+```
+
+Which you can then parse and host w/ 
+[Apiary.io](http://docs.testingit.apiary.io/#): 
+![screenshot](/postimages/advent-2015/test2doc-widgets-api.jpg)
+
+Or use a custom parser and host yourself.
+
+## Getting started with test2doc
+A big goal of mine as I was writing this project was to limit the 
+amount of additional code needed to get test2doc up and running.
+
+Requirements:
+
+1. You must have a `TestMain` for each of the packages you wish to 
+document.
+2. All of the tests in your package must share a single `test.Server` 
+instance (from test2doc/test package)
```

Comment:
I'd like to see a link to https://godoc.org/github.com/adams-sarah/test2doc/test#Server or https://godoc.org/github.com/adams-sarah/test2doc/test